### PR TITLE
vm interpreter

### PR DESCRIPTION
**Pull request by turuslan**
### Description of the Change
VM interpreter (`computeTipSetState` from lotus).
Better `Runtime`.

### Benefits
`computeTipSetState` is required for block production and chain manager.
`Runtime` works as lotus.

### Possible Drawbacks 
Some error checks may be missing.
`StateTree` does not use cache yet.

### CODE REVIEW

File: core/vm/actor/builtin/miner/miner_actor.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    std::vector<uint64_t> deal_ids;
+  };
+
+  struct PreCommittedSector {
+    SectorPreCommitInfo info;
+    uint64_t received_epoch;
+  };
+
+  struct MinerActorState {
+    std::map<uint64_t, PreCommittedSector> precommitted_sectors;
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/builtin/miner/miner_actor.hpp]: AMT Multimap?

File: core/vm/runtime/runtime_types.hpp [hunk 2]
```diff
@@ -38,11 +39,18 @@ namespace fc::vm::runtime {
    * Result of message execution
    */
   struct MessageReceipt {
-    ExitCode exit_code;
+    uint8_t exit_code;
```
Comment (Alexey-N-Chernyshov) [inline core/vm/runtime/runtime_types.hpp]: mb change ExitCode?

File: test/core/vm/runtime/runtime_test.cpp [hunk 3]
```diff
@@ -180,25 +180,16 @@ TEST_F(RuntimeTest, send) {
...
-      .WillOnce(testing::Return(fc::outcome::success(to_actor)));
+      .Times(2)
+      .WillRepeatedly(testing::Return(fc::outcome::success(to_actor)));
   EXPECT_CALL(*invoker_, invoke(Eq(to_actor), _, Eq(method), Eq(params)))
       .WillOnce(testing::Return(fc::outcome::success(res)));
-  EXPECT_CALL(*state_tree_, get(Eq(block_miner_)))
-      .WillOnce(testing::Return(fc::outcome::success(miner_actor)));
   EXPECT_CALL(*state_tree_, set(_, _))
-      .Times(3)
+      .Times(1)
```
Comment (Alexey-N-Chernyshov) [inline test/core/vm/runtime/runtime_test.cpp]: WillOnce

File: core/codec/cbor/cbor_decode_stream.cpp [hunk 4]
```diff
@@ -17,15 +17,11 @@ namespace fc::codec::cbor {
...
-  CborDecodeStream &CborDecodeStream::operator>>(std::vector<uint8_t> &bytes) {
-    if (!cbor_value_is_byte_string(&value_)) {
-      outcome::raise(CborDecodeError::WRONG_TYPE);
-    }
-    size_t size;
-    if (CborNoError != cbor_value_get_string_length(&value_, &size)) {
-      outcome::raise(CborDecodeError::INVALID_CBOR);
+  CborDecodeStream &CborDecodeStream::operator>>(gsl::span<uint8_t> bytes) {
+    auto size = bytesLength();
+    if (static_cast<size_t>(bytes.size()) != size) {
```
Comment (masterjedy) [inline core/codec/cbor/cbor_decode_stream.cpp]: is it ok that I should somehow know the size of vector I'm going to decode?

Comment (masterjedy) [reply core/codec/cbor/cbor_decode_stream.cpp]: It possibly can broke down current functionality.

File: core/primitives/rle_bitset/rle_bitset.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    return s << codec::rle::encode(set);
+  }
+
+  template <class Stream,
+            typename = std::enable_if_t<
+                std::remove_reference_t<Stream>::is_cbor_decoder_stream>>
+  Stream &operator>>(Stream &&s, RleBitset &set) {
+    std::vector<uint8_t> rle;
+    s >> rle;
+    auto decoded = codec::rle::decode<RleBitset::value_type>(rle);
```
Comment (masterjedy) [inline core/primitives/rle_bitset/rle_bitset.hpp]: we've already made some number of convenient macro definitions like OUTCOME_TRY(), EXPECT_OUTCOME_TRUE/FALSE. Something like OUTCOME_EXCEPT() implementing this behavior would be useful. There are places like this where we need to return outcome error in form of exception.

File: core/vm/interpreter/interpreter.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+  outcome::result<Result> interpret(const std::shared_ptr<IpfsDatastore> &ipld, const Tipset &tipset) {
+    if (hasDuplicateMiners(tipset.blks)) {
+      return InterpreterError::DUPLICATE_MINER;
+    }
+
+    auto state_tree = std::make_shared<StateTreeImpl>(ipld, tipset.getParentStateRoot());
+    // TODO(turuslan): FIL-146 randomness from tipset
+    std::shared_ptr<RandomnessProvider> randomness;
+    // TODO(turuslan): indices
```
Comment (masterjedy) [inline core/vm/interpreter/interpreter.cpp]: pls explain

File: core/vm/interpreter/interpreter.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    std::shared_ptr<Indices> indices;
+    auto env = std::make_shared<Env>(randomness, state_tree, indices, std::make_shared<InvokerImpl>(), tipset.height, Address{});
```
Comment (masterjedy) [inline core/vm/interpreter/interpreter.cpp]: clang-format it pls

File: core/codec/cbor/cbor_decode_stream.cpp (hunk 4 above)
Comment (turuslan) [reply core/codec/cbor/cbor_decode_stream.cpp]: There is separate function for vector

File: core/primitives/rle_bitset/rle_bitset.hpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  struct RleBitset : public std::set<uint64_t> {
+    using set::set;
+
+    inline RleBitset(set &&s) : set{s} {}
+  };
+
+  template <class Stream,
+            typename = std::enable_if_t<
+                std::remove_reference_t<Stream>::is_cbor_encoder_stream>>
+  Stream &operator<<(Stream &&s, const RleBitset &set) {
```
Comment (Alexey-N-Chernyshov) [inline core/primitives/rle_bitset/rle_bitset.hpp]: Use CBOR_ENCODE from stream_annotations.

File: core/vm/actor/actor.hpp [hunk 9]
```diff
@@ -34,6 +34,10 @@ namespace fc::vm::actor {
       return method_number == other.method_number;
     }
 
+    inline bool operator!=(const MethodNumber &other) const {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/actor.hpp]: Consider totally_ordered from https://www.boost.org/doc/libs/1_65_0/libs/utility/operators.htm.

File: core/vm/interpreter/interpreter.cpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    for (auto &block : blocks) {
+      if (!set.insert(block.miner).second) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  outcome::result<Address> getMinerOwner(StateTreeImpl &state_tree,
+                                         Address miner) {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/interpreter/interpreter.cpp]: const Address &

File: core/vm/interpreter/interpreter.cpp [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      OUTCOME_TRY(state_tree->set(miner_owner, miner_owner_actor));
+
+      OUTCOME_TRY(receipt,
+                  env->applyMessage(UnsignedMessage{
+                      block.miner,
+                      kSystemActorAddress,
+                      system_actor.nonce,
+                      0,
+                      0,
+                      -1,
```
Comment (Alexey-N-Chernyshov) [inline core/vm/interpreter/interpreter.cpp]: Please, define constants for numbers.

Comment (masterjedy) [reply core/vm/interpreter/interpreter.cpp]: at least, add comments for 1/0/-1

**Pull request merged**